src/coreclr/gc/gc.cpp

```diff
@@ -20530,6 +20555,61 @@ void gc_heap::check_and_set_no_gc_oom()
 #endif //MULTIPLE_HEAPS
 }
 
+#ifdef USE_REGIONS
+bool gc_heap::extend_soh_for_no_gc()
```

Comment:
could we write this in a less repeated way? something like (please do verify this is correct, I just wrote it quickly) -

```cplusplus
bool gc_heap::extend_soh_for_no_gc()
{
    size_t required = soh_allocation_no_gc;
    heap_segment* region = ephemeral_heap_segment;
    
    while (true)
    {
        uint8_t* allocated = (region == ephemeral_heap_segment) ? 
                             alloc_allocated : 
                             heap_segment_allocated (region);
        size_t available = heap_segment_reserved (region) - allocated;
        size_t commit = min (available, required);

        if (grow_heap_segment (region, allocated + commit))
        {
            required -= commit;
            if (required == 0)
            {
                break;
            }

            region = heap_segment_next (region);
            if ((region == nullptr) && !(region = get_new_region (0)))
            {
                break;
            }
        }
        else
        {
            break;
        }
    }

    return (required == 0);
}
```